### ISSUE TITLE
Remembering the last played song and the queue

### ISSUE THREAD

**Issue by BluePixel4k** [vfsfitvnm/ViMusic#35]
### Feature description

If you close the app and restart it, the last played song is no longer displayed at the bottom.

![last played song](https://user-images.githubusercontent.com/77979781/173181415-e7ced7d3-16ad-46a4-9750-af12b78833b5.png)

### Why do you want this feature?

It would be handy if the app would remember the last played song and the queue.

**Comment by vfsfitvnm** [comment]
> if the app would remember the last played song and the queue

Unfortunately it doesn't look like something I'm going to implement - it is more challenging than it looks. Also, I don't know where I could save this information without soiling the relational (:rocket:) DB.

**Comment by BluePixel4k** [comment]
Oh okay, that is a pity to hear. :(
In fact, this is a very handy feature that exists in most other music apps.

Would it possibly relieve the database if only the titles are stored and not additionally the current position in the song?
So that at a restart of the app the song would start at minute 00:00?

**Comment by vfsfitvnm** [comment]
Hmm, I can probably do something

**Comment by BluePixel4k** [comment]
Thx, that would be awesome!

**Comment by BluePixel4k** [comment]
Hi @vfsfitvnm, does your concept https://github.com/vfsfitvnm/ViMusic/commit/e8e69549c67b3a5715e2d94e846d6dfb10b75a19 work or does it not work as desired? :)

**Comment by vfsfitvnm** [comment]
@BluePixel4k It works but it's definitely pure garbage for a couple of reasons:
- You can't safely parcelize a `Bundle` (a `MediaItem`) - but ok this problem can be solved;
- I don't know **when** I should save the player state, e.g. `PlayerService::onDestroy` is not called when the user closes the app and clears the notification (or viceversa). I don't think this is a `androidx.media3` problem.

The last one leads to a path that I honestly I don't want to investigate...

**Comment by Jak2k** [comment]
Maybe just save when starting a new song or playlist.

**Comment by vfsfitvnm** [comment]
I had to drop `android.media3:session` (https://github.com/vfsfitvnm/ViMusic/commit/7e6b5747a2f4118fa14e33a7485b61f7b42ede5a) because it had too many subtle bugs among other things and I couldn't wait 4 months for another release. I also saved 40 lines of code... This says a lot.

`PlayerService::onDestroy` is now called as expected.

**Comment by vfsfitvnm** [comment]
Ah, I forgot to say it's already implemented (8f5bc5e90e0a62afc1d950f6573270440e903987) and will be available within the next release.

**Comment by vfsfitvnm** [comment]
For those who are interested, you can try the debug build [here](https://github.com/vfsfitvnm/ViMusic/actions/runs/2612270855). It will install another APK (named "Debug"), so you don't have to worry about the release build.

An important drawback when enabling the persistent queue feature is the impossibility to dismiss the notification (even if paused) when the activity is in your recent tasks. If you close the app and remove it from the recents, then you will be able to dismiss the paused notification. However, (I think) Android 11 and 12 are not "affected" by this.

**Comment by BluePixel4k** [comment]
That's great!
But how exactly can I download the apk there? I don't see a way to do that below.

**Comment by vfsfitvnm** [comment]
It should be at the bottom of the page. Here's the link https://github.com/vfsfitvnm/ViMusic/suites/7212834982/artifacts/288766062

**Comment by BluePixel4k** [comment]
Strange, I can't download it from my cell phone, but I can download it from my PC.

**Comment by vfsfitvnm** [comment]
Did you manage to install it?

I'm closing this issue as the feature is now implemented, just hold on for the next release!

**Comment by BluePixel4k** [comment]
Yes, the install has worked.
I have already tested it, but unfortunately the version is pretty laggy on my phone (Pixel 5). Do you have any idea why this is?

Generally, the saving of the queue works fine, thank you!
I only noticed that the queue is lost when the app is closed by the OS or by a force stop.

**Comment by BluePixel4k** [comment]
> Don't worry, a debug apk is slow because it serves for testing purposes.

Ah, so I wasn't aware of that.

> Yep, that's intended. If the OS kills the app, then none of its code will be execute anymore. However, this should only happen when you force stop the app, which is not the way you close an application normally.

Ok, unfortunately on my phone and others that don't have as much processor power, unused apps are quickly killed in the background.

Could you possibly add the event that the queue is saved when you pause the song?

**Comment by BluePixel4k** [comment]
> Thanks for reporting. I already noticed that and tried to fix it here (https://github.com/vfsfitvnm/ViMusic/commit/7e42820e7c9897b98dc67cd61133e2ccd4b953fe).

Ah, that's great!

> Was the screen locked?

Unfortunately, I don't know anymore, but I think that he was locked in between.

> Maybe I should lower the interval to e.g. 15 seconds.

This is an idea, does this restart take a lot of power?
Otherwise, you could set the interval even lower.

The cross is also an idea, I had seen this before in a similar way in another app. But of course it is not as convenient for the user.
I would suggest to try it for the first time with the more frequent service restart.

**Comment by vfsfitvnm** [comment]
@BluePixel4k 
I tried lowering the interval to 10s. 
Unfortunately, it still gets killed.

There are two other solutions:
1- Host the player service in a separate process;
2- Set the `targetSdk` to 23 (which is 7 years old..., won't do due to high risk of undesired behaviours and security concerns).

I have also tried the close button solution we discussed already, but it's ugly...

The proper:tm: solution is hosting the player service in a separate process - this is how `androix.media3` does it (in fact, correct me if I'm wrong, ViMusic `v0.1.7` doesn't have this problem, i.e. the notification can stay paused endlessly) - but `androix.media3` has other big problems.

Also, given that doing IPC requires marshalling and stuff, I don't feel like I want to implement it. 

All this to play music...

**Comment by vfsfitvnm** [comment]
Uhm, nevermind - I may have find a solution.

**Comment by vfsfitvnm** [comment]
I hope this is my last message lol. The solution I tried worked good, but sadly A12 [doesn't allow](https://developer.android.com/guide/components/foreground-services#background-start-restrictions) what I was doing to keep the service alive.

However, it looks like disabling battery optimizations (against ViMusic) prevents the system to dismiss the notification when the player is paused. At this point, I'll just a show a dialog which prompts the user to whitelist ViMusic, and we should be good.

**Comment by BluePixel4k** [comment]
Its sounds like a good solution 👍🏼.
I tested the last debug apk ([Rewrite PlayerService](https://github.com/vfsfitvnm/ViMusic/actions/runs/2648386240)) and it looks great! 
Triggering the function to add the app to the whitelist is very user-friendly.

Maybe its a little bit to much, but if I grant to whitelist the app and later I remove the app from the whitelist, the app doesn't notify me again to whitelist the app.

I'm excited for the next official release! 🚀

**Comment by vfsfitvnm** [comment]
Good!
However, I realized disabling battery optimization is not enough for some manufacturers, so I had to re-include the hacky solution I found. Due to its nature, it can be enabled only if necessary.

<img src="https://user-images.githubusercontent.com/46219656/178328154-205c7cf7-9398-4f03-aba1-f4ec139fc938.png" width="30%" />

For reference, turning off battery optimizations is enough for AOSP-ish, but not for One UI (Samsung's ROM).

Also, I'm gonna remove the initial popup in favor of a notification dot to make which leads the user to these settings.
